### ISSUE TITLE
[RemoteConfig] Crash if initializing after app start

### ISSUE THREAD

**Issue by tecbot** [firebase/firebase-ios-sdk#3344]
### [REQUIRED] Step 2: Describe your environment

  * Xcode version: 10.2.1
  * Firebase SDK version: 6.4.0
  * Firebase Component: RemoteConfig
  * Component version: 4.2.0

### [REQUIRED] Step 3: Describe the problem

Stack trace:
```
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000fca8c6b30
-[RCNConfigExperiment initWithDBManager:]

Crashed: com.apple.main-thread
0  libobjc.A.dylib                0x180d87530 objc_msgSend + 16
1  Runner                         0x10284c130 -[RCNConfigExperiment initWithDBManager:] + 4376871216
2  Runner                         0x102841920 -[FIRRemoteConfig initWithAppName:FIROptions:namespace:DBManager:configContent:] + 4376828192
3  Runner                         0x1028450e8 -[FIRRemoteConfigComponent remoteConfigForNamespace:] + 4376842472
4  Runner                         0x102841660 +[FIRRemoteConfig remoteConfigWithFIRNamespace:app:] + 4376827488
5  Runner                         0x102841720 +[FIRRemoteConfig remoteConfig] + 4376827680
6  Runner                         0x102905bec -[FirebaseRemoteConfigPlugin handleMethodCall:result:] + 39 (FirebaseRemoteConfigPlugin.m:39)
7  Flutter                        0x102f55ba8 (Missing)
8  Flutter                        0x102ef96a4 (Missing)
9  Flutter                        0x102f48ad0 (Missing)
10 Flutter                        0x102f04534 (Missing)
11 Flutter                        0x102f07cc8 (Missing)
12 CoreFoundation                 0x181b27650 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 28
13 CoreFoundation                 0x181b27380 __CFRunLoopDoTimer + 864
14 CoreFoundation                 0x181b26bb4 __CFRunLoopDoTimers + 248
15 CoreFoundation                 0x181b21b04 __CFRunLoopRun + 1844
16 CoreFoundation                 0x181b210b0 CFRunLoopRunSpecific + 436
17 GraphicsServices               0x183d2179c GSEventRunModal + 104
18 UIKitCore                      0x1ae38d978 UIApplicationMain + 212
19 Runner                         0x102601d48 main + 5 (AppDelegate.swift:5)
20 libdyld.dylib                  0x1815e68e0 start + 4
```
#### Steps to reproduce:

After upgrading Firebase dependencies from 6.3.0 to 6.4.0 and RemoteConfig from 4.1.0 to 4.2.0 we are getting crashes directly on app start. If the crash happen the next app starts are working. After some time (~1 hour) the crash happen again. We are using Firebase in a Flutter project.

#### Relevant Code:

Entry point: https://github.com/flutter/plugins/blob/master/packages/firebase_remote_config/ios/Classes/FirebaseRemoteConfigPlugin.m#L39

**Comment by ryanwilson** [comment]
Hi @tecbot - looking into this right away. Some quick questions as I start investigating:

- Does this happen on any specific iOS version, or all of them?
- Does it happen on both device and simulator?
- Can you enable Zombies in your application and see where it's failing? It looks like it's trying to access an already deallocated instance of something (based on `KERN_INVALID_ADDRESS`)

Thanks, and sorry for the troubles!

**Comment by ryanwilson** [comment]
I haven't been able to reproduce this using a non-Flutter app, looking at setting up Flutter in order to try it out. If you have a plain reproduction (without `GoogleService-Info.plist`) you could share, it'd be very helpful 🙂

**Comment by mouness2020** [comment]
I have 2 crashes in same class in my apps :
first one
RCNConfigExperiment initWithDBManager 
Fatal Exception: NSInvalidArgumentException-[__NSCFString setTimeZone:]: unrecognized selector sent to instance 0x2800f36f0
--
second one
-[RCNConfigExperiment initWithDBManager:]
Crashed: com.google.perf.FPREventsQueue
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x7265567733237863

[com.test.app_issue_crash_f34c5ae3e6924654a8390f7336a46fd3_DNE_0_v2.txt](https://github.com/firebase/firebase-ios-sdk/files/3388517/com.test.app_issue_crash_f34c5ae3e6924654a8390f7336a46fd3_DNE_0_v2.txt)
[com.test.app_issue_crash_ed38c08906b044109d994521a23efeb5_DNE_0_v2.txt](https://github.com/firebase/firebase-ios-sdk/files/3388518/com.test.app_issue_crash_ed38c08906b044109d994521a23efeb5_DNE_0_v2.txt)

**Comment by paulb777** [comment]
@mouness2020 Thanks for the additional detail.

@ryanwilson, @dmandar I suspect a thread safety issue with the static variable assigned the NSDateFormatter. One thread is creating another instance while another thread is sending the `setTimeZone:` message.

**Comment by mouness2020** [comment]
new crashes happen in version 6.4.0
<img width="1333" alt="Screenshot at Jul 14 06-20-58" src="https://user-images.githubusercontent.com/4884141/61178857-9fef1480-a5ff-11e9-8461-132cf99f8271.png">
[com.newline.syarah_issue_crash_80e7ff5eea8b4185992b08670574f421_DNE_0_v2.txt](https://github.com/firebase/firebase-ios-sdk/files/3389426/com.newline.syarah_issue_crash_80e7ff5eea8b4185992b08670574f421_DNE_0_v2.txt)
[com.newline.syarah_issue_crash_6650134ac76d48528fa3660b4e8d130b_DNE_0_v2.txt](https://github.com/firebase/firebase-ios-sdk/files/3389427/com.newline.syarah_issue_crash_6650134ac76d48528fa3660b4e8d130b_DNE_0_v2.txt)
[com.newline.syarah_issue_crash_af4f450837314836b968679b66abea0d_DNE_0_v2.txt](https://github.com/firebase/firebase-ios-sdk/files/3389428/com.newline.syarah_issue_crash_af4f450837314836b968679b66abea0d_DNE_0_v2.txt)
[com.newline.syarah_issue_crash_debb40baeb19479188b8c0e74cf2316b_DNE_0_v2.txt](https://github.com/firebase/firebase-ios-sdk/files/3389429/com.newline.syarah_issue_crash_debb40baeb19479188b8c0e74cf2316b_DNE_0_v2.txt)

**Comment by ryanwilson** [comment]
Thanks folks for the extra information - I have a pending fix internally at cl/258032937.

We still haven't been able to reproduce this, so if anyone can provide reproduction steps so we can verify the fix works properly, it would be greatly appreciated.

**Comment by dmandar** [comment]
This crash should only happen if you are using FirePerf or if you have multiple active instances of RC (not common) in your app. Please let us know if it s not the case for you.

**Comment by mouness2020** [comment]
Yes that's true i'm using FirePef .. also i have two firebase projects
one for live api and second one for test api but one of them works at time

`   let firebasePlistFileName: String = typeURL != "api" ? "GoogleService-Info-test" : "GoogleService-Info"
        
        if let path = Bundle.main.path(forResource: firebasePlistFileName, ofType: "plist"),let firbaseOptions = FirebaseOptions(contentsOfFile: path){
            
            firbaseOptions.deepLinkURLScheme = self.customURLScheme
            
            FirebaseOptions.defaultOptions()?.deepLinkURLScheme =  self.customURLScheme
            FirebaseConfiguration.shared.setLoggerLevel(.min)
            FirebaseApp.configure(options: firbaseOptions)
            
            
        }
        
        Analytics.setAnalyticsCollectionEnabled(true)
        Performance.sharedInstance().isInstrumentationEnabled = true
        Performance.sharedInstance().isDataCollectionEnabled = true
      `

**Comment by b-onc** [comment]
We also experience this crash, we have a single instance of RC and we do use Firebase Performance. The crash happens half of the times on a clean install. Removing Firebase Performance fixes this issue. In the meantime we downgraded to 6.3.0

**Comment by igiazlas** [comment]
I also have this issue. It happens on fresh installs and old ones.

My configuration on AppDelegate is

```
// Firebase

#if DEVELOP
    let firConfigFilename = "GoogleService-Info-Dev"
#else
    let firConfigFilename = "GoogleService-Info"
#endif

let firConfigPath = Bundle.main.path(forResource: firConfigFilename, ofType: "plist")!

FirebaseApp.configure(options: FirebaseOptions(contentsOfFile: firConfigPath)!)
FirebaseConfiguration.shared.setLoggerLevel(.min)

// Remote Config

let remoteConfig = RemoteConfig.remoteConfig()
remoteConfig.setDefaults(fromPlist: "RemoteConfigDefaults")
remoteConfig.fetchAndActivate()
```

**Comment by ryanwilson** [comment]
@igiazlas thanks for the report. Is reproducible for you on a local test device? If so, what iOS version and what device / simulator are you using?

We have what we believe is a fix going into the next release that's being packaged and tested shortly, hopefully we can find a reliable repro to validate this issue goes away.

**Comment by igiazlas** [comment]
I can reproduce it on an iPhone X running 12.3.1, on a 5s running 12.4 beta but not on a 6s running 11.4.1. Also in all the 12.3.1 simulators I checked. It is consistent if I leave the last 3 lines in AppDelegate. If I remove them, and don't put them anywhere else, it does not crash, at least not for the first 20-30 seconds that I let it run.

**Comment by KevinLaRosa** [comment]
Hi,
Same problem for me on my iPhone X 12.3.1 after a pod update en 6.4.0.
We use Firebase with RemoteConfig and Performance and we have the same setup than @igiazlas except that we set our defaults from a dictionary.

**Comment by igiazlas** [comment]
I also use Firebase Performance as @KevinLaRosa. This is the Firebase values I have on the Podfile.lock

```
- Firebase/Analytics (6.4.0):
    - Firebase/Core
  - Firebase/Auth (6.4.0):
    - Firebase/CoreOnly
    - FirebaseAuth (~> 6.2.0)
  - Firebase/Core (6.4.0):
    - Firebase/CoreOnly
    - FirebaseAnalytics (= 6.0.3)
  - Firebase/CoreOnly (6.4.0):
    - FirebaseCore (= 6.0.4)
  - Firebase/Performance (6.4.0):
    - Firebase/CoreOnly
    - FirebasePerformance (~> 3.1.1)
  - Firebase/RemoteConfig (6.4.0):
    - Firebase/CoreOnly
    - FirebaseRemoteConfig (~> 4.2.0)
  - FirebaseABTesting (3.0.0):
    - FirebaseCore (~> 6.0)
    - Protobuf (~> 3.5)
  - FirebaseAnalytics (6.0.3):
    - FirebaseCore (~> 6.0)
    - FirebaseInstanceID (~> 4.2)
    - GoogleAppMeasurement (= 6.0.3)
    - GoogleUtilities/AppDelegateSwizzler (~> 6.0)
    - GoogleUtilities/MethodSwizzler (~> 6.0)
    - GoogleUtilities/Network (~> 6.0)
    - "GoogleUtilities/NSData+zlib (~> 6.0)"
    - nanopb (~> 0.3)
  - FirebaseAuth (6.2.0):
    - FirebaseAuthInterop (~> 1.0)
    - FirebaseCore (~> 6.0)
    - GoogleUtilities/AppDelegateSwizzler (~> 6.2)
    - GoogleUtilities/Environment (~> 6.2)
    - GTMSessionFetcher/Core (~> 1.1)
  - FirebaseAuthInterop (1.0.0)
  - FirebaseCore (6.0.4):
    - GoogleUtilities/Environment (~> 6.0)
    - GoogleUtilities/Logger (~> 6.0)
  - FirebaseInstanceID (4.2.1):
    - FirebaseCore (~> 6.0)
    - GoogleUtilities/Environment (~> 6.0)
    - GoogleUtilities/UserDefaults (~> 6.0)
  - FirebasePerformance (3.1.1):
    - FirebaseCore (~> 6.0)
    - FirebaseInstanceID (~> 4.2)
    - FirebaseRemoteConfig (~> 4.2)
    - GoogleToolboxForMac/Logger (~> 2.1)
    - "GoogleToolboxForMac/NSData+zlib (~> 2.1)"
    - GoogleUtilities/Environment (~> 6.0)
    - GoogleUtilities/ISASwizzler (~> 6.0)
    - GoogleUtilities/MethodSwizzler (~> 6.0)
    - GTMSessionFetcher/Core (~> 1.1)
    - Protobuf (~> 3.5)
  - FirebaseRemoteConfig (4.2.0):
    - FirebaseABTesting (~> 3.0)
    - FirebaseCore (~> 6.0)
    - FirebaseInstanceID (~> 4.2)
    - GoogleUtilities/Environment (~> 6.0)
    - "GoogleUtilities/NSData+zlib (~> 6.0)"
    - Protobuf (~> 3.5)
  - FirebaseUI/Auth (8.0.4):
    - Firebase/Auth (~> 6.0)
    - GoogleUtilities/UserDefaults
  - FirebaseUI/Facebook (8.0.4):
    - FBSDKLoginKit (~> 5.0)
    - FirebaseUI/Auth
  - FirebaseUI/Google (8.0.4):
    - FirebaseUI/Auth
    - GoogleSignIn (~> 4.0)
```

**Comment by ryanwilson** [comment]
@igiazlas can you verify which simulator versions you're using? As far as I'm aware there are no 12.3.1 simulators, only 12.2.1 and above.

**Comment by KevinLaRosa** [comment]
I've seen that the crash happened when I removed```RemoteConfigSettings(developerModeEnabled: false)```

I have sometime the crash on simulator 12.2
```-[__NSCFString setTimeZone:]: unrecognized selector sent to instance 0x600003615290
2019-07-17 16:34:23.434987+0200 xx 7777:351799] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFString setTimeZone:]: unrecognized selector sent to instance 0x600003615290'```

**Comment by igiazlas** [comment]
@ryanwilson you are right, I meant to say the latest simulators which are 12.2

**Comment by dmandar** [comment]
Based on all your input, we have made a fix for a race that occurs if Perf and RC SDK are both included, which should be out soon. @igiazlas can you also confirm you are also using FirePerf? If not, please confirm if you have the same stack trace.

**Comment by igiazlas** [comment]
@dmandar yes, I also use FirePerf. I've posted my Podfile.lock above.

**Comment by kennic** [comment]
My app also crashes a lot because of this 
<img width="494" alt="Screen Shot 2019-07-18 at 12 41 35 PM" src="https://user-images.githubusercontent.com/5372071/61432081-787aad80-a959-11e9-95a4-54f620d8ee4f.png">
<img width="865" alt="Screen Shot 2019-07-18 at 12 43 45 PM" src="https://user-images.githubusercontent.com/5372071/61432179-cb546500-a959-11e9-9d5f-2216ca27c5e1.png">
<img width="877" alt="Screen Shot 2019-07-18 at 12 44 06 PM" src="https://user-images.githubusercontent.com/5372071/61432180-cb546500-a959-11e9-80a2-231650c9c280.png">

**Comment by bajtyngier** [comment]
Also experiencing the crash at times.
`EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x00000007b9678940`
`-[RCNConfigExperiment initWithDBManager:]`

**Comment by paulb777** [comment]
Firebase 6.5.0 just released with a fix. We were never able to reproduce the issue internally so would appreciate confirmation from those how saw the issue.

**Comment by thomaslange** [comment]
@paulb777 I can confirm the fix. Thanks for your quick help!

**Comment by manishmittal79** [comment]
I was getting the similar crashes on app start during the UITests running on CI using Firebase 6.3. After upgrading to 6.5 today, and running couple of times I can no longer see this crash on startup. Thanks

**Comment by ppamorim** [comment]
Same crash happening on `Firebase (6.7.0)` and `FirebaseCore (6.2.1)`.

**Comment by ryanwilson** [comment]
@ppamorim can you please update to the latest version `Firebase (6.8.1)` and let us know if you're still seeing the crash?

**Comment by ppamorim** [comment]
@ryanwilson I will, thanks.

**Comment by garypez** [comment]
I am seeing this issue with Firebase (6.8.1)
```
Crashed: com.google.GoogleConfigService.FIRRemoteConfig
0  libdispatch.dylib              0x182831bac <redacted> + 452
1  libdispatch.dylib              0x182831724 <redacted> + 144
2  SimpliSafe                     0x100e5947c -[FIRRemoteConfig configValueForKey:namespace:] + 378 (FIRRemoteConfig.m:378)
3  SimpliSafe                     0x100df8c34 -[FPRRemoteConfigFlags sessionSamplingRate] + 4343942196
4  SimpliSafe                     0x100df8294 -[FPRConfigurations sessionsSamplingPercentage] + 4343939732
5  SimpliSafe                     0x100df3188 -[FPRSessionManager isGaugeCollectionEnabledForSessionId:] + 4343918984
6  SimpliSafe                     0x100df3038 -[FPRSessionManager updateSessionId:] + 4343918648
7  SimpliSafe                     0x100df2da0 -[FPRSessionManager initWithNotificationCenter:] + 4343917984
8  SimpliSafe                     0x100df2cfc __35+[FPRSessionManager sharedInstance]_block_invoke + 4343917820
9  libdispatch.dylib              0x182825088 <redacted> + 20
10 libdispatch.dylib              0x182826770 <redacted> + 32
11 SimpliSafe                     0x100df2ca0 +[FPRSessionManager sharedInstance] + 4343917728
12 SimpliSafe                     0x100e03700 -[FPRNetworkTrace start] + 4343985920
13 SimpliSafe                     0x100e0a730 __InstrumentDataTaskWithRequestCompletionHandler_block_invoke + 4344014640
14 SimpliSafe                     0x100e0a6ac __InstrumentDataTaskWithRequestCompletionHandler_block_invoke + 4344014508
15 SimpliSafe                     0x100de2f40 -[FIRInstanceIDCheckinService checkinWithExistingCheckin:completion:] + 189 (FIRInstanceIDCheckinService.m:189)
16 SimpliSafe                     0x100de06bc -[FIRInstanceIDAuthService fetchCheckinInfoWithHandler:] + 188 (FIRInstanceIDAuthService.m:188)
17 SimpliSafe                     0x100dd9d28 -[FIRInstanceID tokenWithAuthorizedEntity:scope:options:handler:] + 368 (FIRInstanceID.m:368)
18 SimpliSafe                     0x100e67528 -[RCNConfigFetch refreshInstanceIDTokenAndFetchCheckInInfoWithCompletionHandler:] + 230 (RCNFetch.m:230)
19 SimpliSafe                     0x100e6718c __74-[RCNConfigFetch fetchAllConfigsWithExpirationDuration:completionHandler:]_block_invoke + 190 (RCNFetch.m:190)
20 libdispatch.dylib              0x182823c2c <redacted> + 32
21 libdispatch.dylib              0x182825088 <redacted> + 20
22 libdispatch.dylib              0x18282b500 <redacted> + 568
23 libdispatch.dylib              0x18282bf2c <redacted> + 368
24 libdispatch.dylib              0x182834f9c <redacted> + 580
25 libsystem_pthread.dylib        0x18288aaa4 _pthread_wqthread + 280
26 libsystem_pthread.dylib        0x182890c7c start_wqthread + 8```

**Comment by morganchen12** [comment]
@dmandar looks like a regression, can you investigate?

**Comment by paulb777** [comment]
@garypez This looks like #3842. Please follow the steps listed there to make sure you have a clean build.

**Comment by paulb777** [comment]
I'll close this one. Please follow up in #3842 if you're unable to resolve.

**Comment by valerianaGit** [comment]
@paulb777 @ryanwilson hi guys. I am trying to link performance today and getting these same crashes on app launch. Will follow #3842 , but I am running iOS 12 still.

**Comment by ryanwilson** [comment]
@valerianaGit please make sure you copy **all** of the frameworks from the Performance folder in the zip file. `FirebaseRemoteConfig.framework` is part of the `Performance` folder and is a required dependency of Performance.

**Comment by valerianaGit** [comment]
@ryanwilson thanks so much for the response. I did copy the entire folder and (after clearing derived data, clean build and restarting my machine twice) I still get the above crash. And it crashes here in `AppDelegate` in `didFinishLaunchingWithOptions` 
<img src="https://user-images.githubusercontent.com/42816552/67109594-757c9c80-f196-11e9-90fe-48b72288f09f.png" height=30px> 
I followed the steps in `getting started with Performance`. 

Is there anything else we might be missing to initialize or anything?
